[Flang][runtime] Distinguish CPU time and elapsed time for cpu_time and system_clock

mjklemm opened the pull request:
The current implementation for `system_clock()` returns the CPU time instead of elapsed wallclock time.

mjklemm committed:
Distinguish CPU time and elapsed time

mjklemm committed:
Rework fallback implementation to use C++11 high-res clock

mjklemm committed:
Make clang-format happy

jeanPerier commented:
I am bit worried about the C++ runtime dependency that this could bring if `std::chrono::high_resolution_clock` is not header only in some C++ standard library. Did check that? Could you add a system_clock test in `test/Runtime/no-cpp-dep.c`?

mjklemm committed:
Add SystemClockCount to the no-cpp-dep.c test

mjklemm committed:
Fix Windows build

mjklemm committed:
Test if system_clock() is returning a meaningful result

ergawy reviewed (commented):
Just one question about the lit test. Thanks @mjklemm.

mjklemm commented on flang/test/Runtime/system_clock.f90:
```diff
@@ -0,0 +1,27 @@
+
+    diff = real(t_end - t_start, kind=real64) / real(rate, kind=real64)
+
+    if (abs(diff - real(delta, kind=real64)) <= epsilon) then
```
Hm. Not sure how to test if time returned by system_clock() is close to the actual time that has passed.

But I agree that the test is flaky. I can remove it, if we feel that we are OK without it.

mjklemm replied:
I have removed that test now.

kiranchandramohan commented:
> I am bit worried about the C++ runtime dependency that this could bring if `std::chrono::high_resolution_clock` is not header only in some C++ standard library. Did check that? Could you add a system_clock test in `test/Runtime/no-cpp-dep.c`?

Yeah, using chrono related stuff has always ended up with some dependency. Most recently it came in as part of Sleep and it was removed in https://github.com/llvm/llvm-project/pull/84911.

mjklemm committed:
Reduce precision requirement for the test

mjklemm committed:
Dump std::chrono and use timespec_get instead

With commit, some functions seem to be very close to each other, so some cleanup might be desired.

mjklemm committed:
Remove constantly failing test

mjklemm committed:
Report proper max for count_max

After the changes made, we do not rely on std::clock_t anymore, so
we can report a much higher limit for the maximum.

luporl commented on flang/runtime/time-intrinsic.cpp:
```diff
@@ -19,6 +19,7 @@
 #include <cstdlib>
 #include <cstring>
 #include <ctime>
+#include <chrono>
```
Is this include still needed?

mjklemm replied:
Nope. It can be removed.  I think the code can be greatly simplified.

If there's consensus amongst the reviewers that the way I did things are fine, I'll refactor the code and remove everything that's no longer needed.

mjklemm committed:
Refactor code a bit to reduce redundancy

mjklemm commented:
Alrighty then.  Let me see if I can remove the `std::chrono` stuff.

> > I am bit worried about the C++ runtime dependency that this could bring if `std::chrono::high_resolution_clock` is not header only in some C++ standard library. Did check that? Could you add a system_clock test in `test/Runtime/no-cpp-dep.c`?
> 
> Yeah, using chrono related stuff has always ended up with some dependency. Most recently it came in as part of Sleep and it was removed in #84911.

We now have two implementations

- one based on `std::timespec_get` as the fallback
- one based on `clock_gettime`

I have also done some mild refactoring of the code to slightly reduce duplication due to my changes.

What do y'all think?

ergawy reviewed (approved):
LGTM, verified that this works properly by comparing the timings to `omp_get_wtime()`.

luporl reviewed (approved):
I'm not familiar with this part of the code, but the changes make sense to me and look good.

mjklemm commented:
@rovka @vdonaldson Would you like to have a quick look at this?

DanielCChen commented:
> Alrighty then. Let me see if I can remove the `std::chrono` stuff.
> 
> > > I am bit worried about the C++ runtime dependency that this could bring if `std::chrono::high_resolution_clock` is not header only in some C++ standard library. Did check that? Could you add a system_clock test in `test/Runtime/no-cpp-dep.c`?
> > 
> > 
> > Yeah, using chrono related stuff has always ended up with some dependency. Most recently it came in as part of Sleep and it was removed in #84911.
> 
> We now have two implementations
> 
> * one based on `std::timespec_get` as the fallback
> * one based on `clock_gettime`
> 
> I have also done some mild refactoring of the code to slightly reduce duplication due to my changes.
> 
> What do y'all think?

Is there a particular reason to go in favor of `std::timespec_get` over POSIX function `clock_gettime`?
 I posted PR #98915 to fix the build breakage on AIX because of it.

mjklemm commented:
> Is there a particular reason to go in favor of `std::timespec_get` over POSIX function `clock_gettime`?
> I posted PR #98915 to fix the build breakage on AIX because of it. @mjklemm

`clock_gettime` is still there.  Before this PR, the code used the wrong entry point and did not return time, but returned CPU time, which was not correct for what the Fortran intrinsic was supposed to return.  I'll comment in your PR also.

DanielCChen commented:
Thanks for the explanation. The reason of Flang build failure on AIX is because macro `TIME_UTC` does not exist on AIX 7.2. The build failed at the compile time with "undefined symbol". The code that references `TIME_UTC` is not "guarded".
